serveradmin/serverdb/query_materializer.py

```diff
@@ -197,62 +196,59 @@ def _add_domain_attribute(self, attribute, servers):
             )
 
     def _add_supernet_attribute(self, attribute: Attribute, servers_in):
-        if attribute.inet_address_family:
-            self._add_supernet_attribute_af(attribute, servers_in)
-        else:
-            self._add_supernet_attribute_intern_ip(attribute, servers_in)
-
-    def _add_supernet_attribute_af(self, attribute, servers_in):
-        supernet_id = ServerInetAttribute.objects.filter(
-            value__net_contains_or_equals=OuterRef("value"),
-            server__servertype_id=attribute.target_servertype_id,
-            attribute__inet_address_family=attribute.inet_address_family,
-        ).values("server__server_id")
-
-        attrs = ServerInetAttribute.objects.filter(
-            server_id__in=[s.server_id for s in servers_in],
-            attribute__inet_address_family=attribute.inet_address_family,
-        ).annotate(
-            supernet_id=Subquery(supernet_id, output_field=AutoField()),
-        )
-
-        supernets = Server.objects.filter(server_id__in=attrs.values("supernet_id"))
-        supernets_by_id = {s.server_id: s for s in supernets}
-
-        for a in attrs:
-            if a.supernet_id in supernets_by_id:
-                self._server_attributes[a.server][attribute] = supernets_by_id[
-                    a.supernet_id
-                ]
+        q = """
+            SELECT
+                net.server_id, net.hostname, net.intern_ip, net.servertype_id,
+                net_attr.inet_address_family, host_attr.inet_address_family,
+                host.server_id AS host_server_id, host_attr.attribute_id AS host_attr_name
+            FROM server AS host
+            JOIN server_inet_attribute AS host_addr ON (host.server_id = host_addr.server_id)
+            JOIN server_inet_attribute AS net_addr ON (host_addr.value <<= net_addr.value AND host_addr.attribute_id = net_addr.attribute_id)
+            JOIN server AS net ON (net_addr.server_id = net.server_id)
+            JOIN attribute AS net_attr ON (net_attr.attribute_id = net_addr.attribute_id)
+            JOIN attribute AS host_attr ON (host_attr.attribute_id = host_addr.attribute_id)
+            WHERE
+                net.servertype_id = %(target_servertype)s
+                AND host.server_id = any(%(hosts)s)
+        """
 
-    def _add_supernet_attribute_intern_ip(self, attribute, servers):
-        servers_in_1, servers_in_2 = itertools.tee(servers)
+        if attribute.inet_address_family:
```

Comment:
In which case could this be blank ?

Reply:
For example `route_network` attribute has no address family. For this supernet the code should check all inet type attributes and ensure that they point to the same supernet. `route_networks` are dual stack.

On the other hand `provider_network_ipv4` and `provider_network_ipv6` have address family. A host can belong to multiple provider networks, like a local subnet matching on `10.0.0.0/8` for Legacy IP and some a IPv6 subnet, like `2001:db8::/32`.